vllm/model_executor/guided_decoding/outlines_logits_processors.py

```diff
@@ -7,59 +7,94 @@
-import copy
+
+import importlib.metadata
 import json
-from collections import defaultdict
-from functools import lru_cache
-from typing import Callable, DefaultDict, Dict, List, Optional, Union
+import os
+import re
+from typing import Dict, List, Optional, Union
 
-import numpy as np
+import diskcache as dc
 import torch
-from outlines import grammars
-from outlines.caching import cache, disable_cache
-from outlines.fsm.guide import (CFGGuide, CFGState, Generate, Guide,
-                                RegexGuide, Write)
-from outlines.fsm.parsing import PartialLark
-from outlines_core.fsm.json_schema import build_regex_from_schema
+from cachetools import LRUCache
+from diskcache import Cache
+from outlines_core import Guide, Index, Vocabulary
+from outlines_core.json_schema import build_regex_from_schema
+from outlines_core.kernels.torch import (_apply_token_bitmask_inplace_kernel,
+                                         allocate_token_bitmask)
 from pydantic import BaseModel
 from transformers import PreTrainedTokenizerBase
+from transformers.models.gpt2.tokenization_gpt2 import bytes_to_unicode
 
 import vllm.envs as envs
 from vllm.logger import init_logger
-from vllm.platforms import current_platform
-from vllm.reasoning import ReasoningParser
+from vllm.model_executor.guided_decoding.reasoner import Reasoner
+from vllm.transformers_utils.tokenizer import AnyTokenizer
 
 logger = init_logger(__name__)
 
+CACHE = None
+
+
+def get_disk_cache() -> dc.Cache:
+    """Get the context object that contains previously-computed return values"""
+    outlines_cache_dir = os.getenv("OUTLINES_CACHE_DIR")
+    xdg_cache_home = os.getenv("XDG_CACHE_HOME")
+    home_dir = os.path.expanduser("~")
+    outlines_version = importlib.metadata.version("outlines_core")
+
+    if outlines_cache_dir:
+        # OUTLINES_CACHE_DIR takes precendence
+        cache_dir = outlines_cache_dir
+    elif xdg_cache_home:
+        cache_dir = os.path.join(xdg_cache_home, ".cache", "outlines")
+    elif home_dir != "/":
+        cache_dir = os.path.join(home_dir, ".cache", "outlines")
+    else:
+        import tempfile
+
+        # home_dir may be / inside a docker container without existing user
+        tempdir = tempfile.gettempdir()
+        cache_dir = os.path.join(tempdir, ".cache", "outlines")
+
+    memory = Cache(cache_dir, eviction_policy="none", cull_limit=0)
+
+    # Ensure if a version upgrade occurs, old cache is pruned
+    cached_version = memory.get('__version__', None)
+    if cached_version != outlines_version:
+        memory.clear()
+    memory.set('__version__', outlines_version)
+
+    return memory
+
+
 if envs.VLLM_V0_USE_OUTLINES_CACHE:
     logger.warning("Enabling outlines cache. This is an unbounded on-disk "
                    "cache. It may consume a lot of disk space and should "
                    "not be used with untrusted clients.")
+    CACHE = get_disk_cache()
 else:
-    disable_cache()
+    CACHE = LRUCache(maxsize=128)
```

Comment:
this means that this will get called everytime this modules is imported.

i prefer not to have this side effect here. Note that if the worker process get forked, this will also get called multiple times.